### ISSUE TITLE
Tracking down PMIX_ERR_PACK_MISMATCH problem

### ISSUE THREAD

**Issue by krempel-pt** [openpmix/openpmix#3204]
### What version of the PMIx Reference Library are you using?
5.0.1

### Describe how PMIx was installed
release tarball

Configure command line: '--build=x86_64-redhat-linux-gnu'
                          '--host=x86_64-redhat-linux-gnu'
                          '--program-prefix=' '--disable-dependency-tracking'
                          '--prefix=/usr' '--exec-prefix=/usr'
                          '--bindir=/usr/bin' '--sbindir=/usr/sbin'
                          '--sysconfdir=/etc' '--datadir=/usr/share'
                          '--includedir=/usr/include' '--libdir=/usr/lib64'
                          '--libexecdir=/usr/libexec' '--localstatedir=/var'
                          '--runstatedir=/run' '--sharedstatedir=/var/lib'
                          '--mandir=/usr/share/man'
                          '--infodir=/usr/share/info' '--prefix=/usr'
                          '--sysconfdir=/etc/pmix' '--disable-static'
                          '--disable-silent-rules' '--enable-shared'
                          '--with-munge'

### Please describe the system on which you are running

* Operating system/version: Fedora release 38
* Computer hardware: VM qemu/KVM x86_64
* Network type: virtio

-----------------------------

## Details of the problem

I am trying to track down the problem already mentioned by @sonjahapp in https://github.com/openpmix/openpmix/issues/3140#issuecomment-1709704504_

On some systems sometimes without willfully changing something, we see the following behavior when calling `PMIx_Init()` in a process set up by our ParaStation process management which provides a PMIx server implementation called `pspmix` using the openpmix server interface:

```
$ mpiexec --pmix -np 1 -H node01 Projekte/psmgmt/tests/PMIx/getSizes
[node01:24170] PMIX ERROR: PMIX_ERR_PACK_MISMATCH in file ../../src/client/pmix_client.c at line 278
[pspmix_0x00005e66[0:24166]:0]: PMIx_Init failed: PMIX_ERROR
PSIlogger: Child with rank 0 exited with status 1.
```

In the syslog we can see, that the pmix server library called the `client_connected2` callback before, so the server does not become aware of the problem somehow, it manifests only in the client.

The problem does not always show, but mostly if it shows, then it is reproducible. I currently have this situation again and want to use it to understand what is going on and hopefully find a way to prevent that.

Unfortunately, I have no idea yet what this `PMIX_ERR_PACK_MISMATCH` want to tell me exactly and what problem triggers it. As we do only see it with ParaStation but never with PRRTE, my first guess is that we do something differently when using the server library, maybe setting up the namespace or something.

Currently, I am a bit lost in how to start best. Do you perhaps have a hint for me?

Could you please elaborate a little bit on what `PMIX_ERR_PACK_MISMATCH` actually means here or how I can track that down?
To me it looks like there are some data received from the server and inside there should be some namespace data and for some reason it is not possible to unpack them. Probably since they are corrupted or in an unexpected format?

Is there a debugging option that makes prte show the info fields that are used to create the namespace? Might be helpful to just compare what we are doing differently to get the trigger "from outside"?

Many thanks in advance.

**Comment by rhc54** [comment]
The "pack mismatch" error means that the data type we are trying to unpack doesn't match the data type in the buffer. On that line, we are trying to unpack a string, and for some reason PMIx believes that some other type was packed.

This lies outside the purview of your server host - the corresponding pack operation is done solely inside the PMIx server library. The usual cause is that one side (e.g., your host server) was configured with `--enable-debug` and the other side (e.g., the client) was not. When configured without debug enabled, PMIx doesn't store an explicit data type in the packed buffer. So if a debug-enabled code attempts to unpack it, it gets an unexpected value when trying to read the type.

We are supposed to detect that situation and adjust accordingly, so it would be a bug if we fail to do so. I would therefore suggest first checking to see if the client and server are using the same library (both version and build so you know they have the same configure line).

**Comment by rhc54** [comment]
Sigh - nerts! I'll have to see if it works with v4.2 on both sides. Could be something got lost in the v5 family. Will try to resolve this soon.

**Comment by rhc54** [comment]
Appears to work with v4.2 on both sides (one debug and the other not). Checking v5 to see if I can reproduce it.

**Comment by krempel-pt** [comment]
Thanks for looking into that.

I observe the issue with the combination described above:
- Server: current master (v1.1.3-3934-ga84d8fb9) from git and debugging on
- Client: release 5.0.1 from tarball without debugging

Tomorrow, I will test with different combinations and report the results. Do you have special requests on meaningful combinations?

**Comment by rhc54** [comment]
Well, we know those two won't work as-is (see my note to the PMIx mailing list). You'd need to put `PMIX_MCA_gds=hash` in the environment for both client and server. Not sure if that's what is behind what you are seeing. Note that I avoided the cross-version breakage in my testing.

**Comment by krempel-pt** [comment]
Setting `PMIX_MCA_gds=hash` on both sides seem to fix the `PMIX_ERR_PACK_MISMATCH` error in my described test case. :+1: 

Interestingly, if I am using the same build of PMIx 5.0.1 tarball release (see the starting post of this issue for exact configure line reported by `pmix_info`) on the server and the client side without setting  `PMIX_MCA_gds=hash` anywhere, I get a new error message:
```
[node01:25301] PMIX ERROR: PMIX_ERR_OUT_OF_RESOURCE in file client/pmix_client.c at line 278
[pspmix_0x000062d1[0:25297]:0]: PMIx_Init failed: PMIX_ERROR
```
If I am setting `PMIX_MCA_gds=hash` in the very same scenario only on the client side, the error disappears and everything seem to work again.

Does that fit into the picture?

**Comment by samuelkgutierrez** [comment]
I'll add this to my list.

**Comment by rhc54** [comment]
I think we have this resolved now on master and queued up for PMIx v5.0.2, so I'm going to close this for now.
